Branch-SDK/src/io/branch/indexing/ContentDiscoverer.java

```diff
@@ -0,0 +1,256 @@
+    private String referredUrl_; // The url which opened this app session
+
+    private static final String TIME_STAMP_KEY = "ts";
+    private static final String NAV_PATH_KEY = "n";
+    private static final String REFERRAL_LINK_KEY = "rl";
+    private static final String CONTENT_LINK_KEY = "cl";
+    private static final String CONTENT_META_DATA_KEY = "cm";
+    private static final String VIEW_KEY = "v";
+    private static final String CONTENT_DATA_KEY = "cd";
+    private static final String CONTENT_KEYS_KEY = "ck";
+    private static final String PACKAGE_NAME_KEY = "p";
+    private static final String ENTITIES_KEY = "e";
+
+    private final HashHelper hashHelper_;
+    private ContentDiscoveryManifest cdManifest_;
+
+    public static ContentDiscoverer getInstance() {
+        if (thisInstance_ == null) {
+            thisInstance_ = new ContentDiscoverer();
+        }
+        return thisInstance_;
+    }
+
+    private ContentDiscoverer() {
+        handler_ = new Handler();
+        hashHelper_ = new HashHelper();
+    }
+
+    private int discoveredViewInThisSession_ = 0; // Denote the number  of views discovered in this session
+    private ArrayList<String> discoveredViewList_ = new ArrayList<>(); // List for saving already discovered views path
+
+    //------------------------- Public methods---------------------------------//
+
+    public void discoverContent(final Activity activity, String referredUrl) {
+        cdManifest_ = ContentDiscoveryManifest.getInstance(activity);
+        referredUrl_ = referredUrl;
+
+        //Scan for content only if the app is started by  a link click or if the content path for this view is already cached
+        ContentDiscoveryManifest.CDPathProperties pathProperties = cdManifest_.getCDPathProperties(activity);
+
+        if (pathProperties != null) { // Check if view is available in CD manifest
+            // Discover content only if element array is not empty json array
+            if (!pathProperties.isSkipContentDiscovery()) {
+                discoverContent(activity);
+            }
+        } else if (!TextUtils.isEmpty(referredUrl_)) {
+            discoverContent(activity);
+        }
+    }
+
+    private void discoverContent(Activity activity) {
+        if (discoveredViewList_.size() < cdManifest_.getMaxViewHistorySize()) { // check if max discovery views reached
+            handler_.removeCallbacks(readContentRunnable);
+            lastActivityReference_ = new WeakReference<>(activity);
+            handler_.postDelayed(readContentRunnable, VIEW_SETTLE_TIME);
+        }
+    }
+
+
+    public void onActivityStopped(Activity activity) {
+        if (lastActivityReference_ != null && lastActivityReference_.get() != null
+                && lastActivityReference_.get().getClass().getName().equals(activity.getClass().getName())) {
+            handler_.removeCallbacks(readContentRunnable);
+            lastActivityReference_ = null;
```

Comment:
The runnable happens in the background - right? It would seems it's possible for the runnable to start executing, pass through line 119, then this code to be executed resulting in you calling `get()` on null in 120.